File: src/main/java/com/nickcoblentz/montoya/settings/ExtensionSetting.java
package com.nickcoblentz.montoya.settings;

import burp.api.montoya.MontoyaApi;
import com.nickcoblentz.montoya.MontoyaLogger;

public class ExtensionSetting implements IExtensionSetting {
    private final String DefaultValue;
    private final String StorageKey;
    private final String Name;
    private final MontoyaApi Api;
    private final ExtensionSettingSaveLocation DefaultSaveLocation;
    private final MontoyaLogger Logger;
    private String CurrentValue;



    public ExtensionSetting(MontoyaApi api, String name, String storageKey, String defaultValue, ExtensionSettingSaveLocation defaultSaveLocation) {
        Logger = new MontoyaLogger(api);
        Api=api;
        Name=name;
        StorageKey=storageKey;
        DefaultValue=defaultValue;
        DefaultSaveLocation=defaultSaveLocation;
        load();

    }

    @Override
    public String getCurrentValue() {
        return CurrentValue==null || CurrentValue.isEmpty() ? DefaultValue : CurrentValue;
    }

    @Override
    public void setCurrentValue(String currentValue) {
        if(currentValue!=null)
            CurrentValue = currentValue;
    }

    @Override
    public void saveToProject()
    {
        Api.persistence().extensionData().setString(StorageKey,getCurrentValue());
    }

    @Override
    public void loadFromProject()
    {
        setCurrentValue(Api.persistence().extensionData().getString(StorageKey));
    }

    @Override
    public void saveToPreferences()
    {
        Api.persistence().preferences().setString(StorageKey,getCurrentValue());
    }

    @Override
    public void loadFromPreferences()
    {
        setCurrentValue(Api.persistence().preferences().getString(StorageKey));
    }

    @Override
    public void save() {
        switch(DefaultSaveLocation) {
            case ExtensionSettingSaveLocation.PREFERENCE -> saveToPreferences();
            case ExtensionSettingSaveLocation.PROJECT -> saveToProject();
            default -> {}
        }
    }

    @Override
    public void load() {
        switch(DefaultSaveLocation) {
            case ExtensionSettingSaveLocation.PREFERENCE -> loadFromPreferences();
            case ExtensionSettingSaveLocation.PROJECT -> loadFromProject();
            default -> {}
        }
    }

    @Override
    public String getName() {
        return Name;
    }

    @Override
    public String getStorageKey() {
        return StorageKey;
    }

    @Override
    public String reset()
    {
        setCurrentValue(DefaultValue);
        return getCurrentValue();
    }
}

File: src/main/java/com/nickcoblentz/montoya/settings/ExtensionSetting.kt
package com.nickcoblentz.montoya.settings

import burp.api.montoya.MontoyaApi
import com.nickcoblentz.montoya.MontoyaLogger

class ExtensionSetting(private val api: MontoyaApi, override val name: String,override val storageKey: String, public val defaultValue: String, public override val defaultSaveLocation: ExtensionSettingSaveLocation) : IExtensionSetting {
    private val Logger = MontoyaLogger(api)

    override var currentValue: String = defaultValue
        get() {
            return if (field == null || field.isEmpty()) defaultValue else field
        }
        set(value) {
            if (value != null) field = value
        }

    init {
        load()
    }

    override fun saveToProject() {
        api.persistence().extensionData().setString(storageKey, currentValue)
    }

    override fun loadFromProject() {
        currentValue = api.persistence().extensionData().getString(storageKey)
    }

    override fun saveToPreferences() {
        api.persistence().preferences().setString(storageKey, currentValue)
    }

    override fun loadFromPreferences() {
        currentValue = api.persistence().preferences().getString(storageKey)
    }

    override fun save() {
        when (defaultSaveLocation) {
            ExtensionSettingSaveLocation.PREFERENCE -> saveToPreferences()
            ExtensionSettingSaveLocation.PROJECT -> saveToProject()
            else -> {}
        }
    }

    override fun load() {
        when (defaultSaveLocation) {
            ExtensionSettingSaveLocation.PREFERENCE -> loadFromPreferences()
            ExtensionSettingSaveLocation.PROJECT -> loadFromProject()
            else -> {}
        }
    }

    override fun reset(): String? {
        currentValue = defaultValue
        return currentValue
    }
}